### PR TITLE
Feature/apl 575 sharding recovery

### PR DESCRIPTION
**Pull request by yuriy-larin-firstbridge**

### CODE REVIEW

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/AplDbVersion.java [hunk 1]
```diff
@@ -688,38 +688,52 @@ protected int update(int nextUpdate) {
                 apply("CREATE INDEX IF NOT EXISTS genesis_public_key_height_idx on genesis_public_key(height)");
             case 251:
                 // SHARDING meta-info inside main database
-                apply("CREATE TABLE IF NOT EXISTS shard (shard_id BIGINT AUTO_INCREMENT NOT NULL, shard_hash VARBINARY not null, shard_state BIGINT)");
+                apply("CREATE TABLE IF NOT EXISTS shard (shard_id BIGINT AUTO_INCREMENT NOT NULL, shard_hash VARBINARY, " +
+                        "shard_height BIGINT not null, shard_state BIGINT)");
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/AplDbVersion.java]: height should be int

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/AplDbVersion.java [hunk 2]
```diff
@@ -688,38 +688,52 @@ protected int update(int nextUpdate) {
...
-            case 258:
-                apply("CREATE UNIQUE INDEX IF NOT EXISTS transaction_index_shard_1_idx ON transaction_shard_index (transaction_id, block_id)");
             case 259:
+                apply("CREATE UNIQUE INDEX IF NOT EXISTS transaction_index_shard_1_idx ON transaction_shard_index (transaction_id, block_id)");
+            case 260:
+                apply("CREATE TABLE IF NOT EXISTS referenced_shard_transaction (db_id BIGINT auto_increment NOT NULL, transaction_id BIGINT NOT NULL, " +
+                        "referenced_transaction_id BIGINT NOT NULL)");
+            case 261:
+                apply("ALTER TABLE referenced_shard_transaction ADD CONSTRAINT IF NOT EXISTS pk_referenced_shard_transaction_db_id PRIMARY KEY(db_id)");
+            case 262:
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/AplDbVersion.java]: Remove referenced shard transaction. This table is not exist after referenced transaction integration

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/AplDbVersion.java [hunk 3]
```diff
@@ -688,38 +688,52 @@ protected int update(int nextUpdate) {
...
-            case 262:
+            case 266:
+                apply("ALTER TABLE shard_recovery ADD CONSTRAINT IF NOT EXISTS shard_recovery_id_state_object_idx unique (shard_recovery_id, state)");
+            case 267:
                 apply("ALTER TABLE genesis_public_key DROP CONSTRAINT IF EXISTS CONSTRAINT_C11");
-            case 263:
-                apply("ALTER TABLE referenced_transaction ADD COLUMN height INT NOT NULL");
-            case 264:
+            case 268:
+                apply("ALTER TABLE IF EXISTS referenced_transaction ADD COLUMN IF NOT EXISTS height INT NOT NULL");
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/AplDbVersion.java]: referenced_transaction table should exist here

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/ShardDataSourceCreateHelper.java [hunk 4]
```diff
@@ -57,9 +57,9 @@ public TransactionalDataSource getShardDb() {
     public ShardDataSourceCreateHelper createUninitializedDataSource() {
         if (shardId == null) {
             try (Connection con = databaseManager.getDataSource().getConnection();
-                 PreparedStatement pstmt = con.prepareStatement("SELECT IFNULL(max(SHARD_ID) + 1, 1) as shard_id FROM shard")) {
+                 PreparedStatement pstmt = con.prepareStatement("SELECT IFNULL(max(SHARD_ID), 0) as shard_id FROM shard")) {
                 try (ResultSet rs = pstmt.executeQuery()) {
-                    while (rs.next()) {
+                    if (rs.next()) {
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/ShardDataSourceCreateHelper.java]: else throw exception

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/DataTransferManagementReceiverImpl.java [hunk 5]
```diff
@@ -417,10 +417,11 @@ public MigrateState addShardInfo(CommandParamInfo paramInfo) {
...
         try (Connection sourceConnect = sourceDataSource.begin();
             PreparedStatement preparedInsertStatement = sourceConnect.prepareStatement(
-                    "insert into SHARD (SHARD_ID, SHARD_HASH, SHARD_STATE) values (?, ?, ?)")) {
-            preparedInsertStatement.setLong(1, createdShardId.get());
-            preparedInsertStatement.setBytes(2, paramInfo.getShardHash());
-            preparedInsertStatement.setInt(3, 100); // 100% full shard is present on current node
+                    "UPDATE SHARD SET SHARD_HASH = ?, SHARD_STATE = ? WHERE SHARD_ID = ?")) {
+            preparedInsertStatement.setBytes(1, paramInfo.getShardHash());
+//            preparedInsertStatement.setLong(1, paramInfo.getSnapshotBlockHeight());
+            preparedInsertStatement.setInt(2, 100); // 100% full shard is present on current node
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/DataTransferManagementReceiverImpl.java]: Move 100 to constants and name it correctly

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/helper/BlockDeleteHelper.java [hunk 6]
```diff
@@ -80,33 +81,53 @@ private boolean handleResultSet(PreparedStatement ps, PaginateResultWrapper pagi
...
                     log.error("Failed Deleting " + currentTableName, e);
                     sourceConnect.rollback();
                     throw e;
                 }
-                rows++;
             }
+        }
+/*
+        try {
+            int[] array = preparedInsertStatement.executeBatch();
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/helper/BlockDeleteHelper.java]: Better move batch delete to another method

File: apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/db/BlockDaoTest.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+import org.jboss.weld.junit5.WeldSetup;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+@EnableWeld
+class BlockDaoTest {
+
+    @RegisterExtension
+    DbExtension extension = new DbExtension(DbTestData.getDbFileProperties(createPath("targetDb").toAbsolutePath().toString()));
```
Comment (AndrewBoyarsky) [inline apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/db/BlockDaoTest.java]: Why we require here db in file?

File: apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/db/BlockDaoTest.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    @AfterEach
+    void tearDown() {
+        extension.getDatabaseManger().shutdown();
+    }
+
+    @Test
+    void countByHeight() {
+        long count = blockDao.getBlockCount(0, 8000);
```
Comment (AndrewBoyarsky) [inline apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/db/BlockDaoTest.java]: Use BlockTestData.Block.getHeight

File: apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/db/ShardDataSourceCreateHelperTest.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+            PropertiesHolder.class, ShardRecoveryDaoJdbcImpl.class, EpochTime.class)
+            .addBeans(MockBean.of(extension.getDatabaseManger(), DatabaseManager.class))
+            .addBeans(MockBean.of(mock(NtpTime.class), NtpTime.class))
+            .build();
+
+    private ShardDataSourceCreateHelper createHelper;
+
+    @AfterAll
+    static void tearDown() {
+        extension.getDatabaseManger().shutdown();
```
Comment (AndrewBoyarsky) [inline apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/db/ShardDataSourceCreateHelperTest.java]: extension will shutdown automatically. Delete this pls

File: apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/db/BlockDaoTest.java [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+            return temporaryFolderExtension.newFolder().toPath().resolve(fileName);
+        }
+        catch (IOException e) {
+            throw new RuntimeException(e.toString(), e);
+        }
+    }
+
+    @AfterEach
+    void tearDown() {
+        extension.getDatabaseManger().shutdown();
```
Comment (AndrewBoyarsky) [inline apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/db/BlockDaoTest.java]: Extension support automatic shutdown

File: apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/db/ShardRecoveryDaoJdbcTest.java [hunk 11]
```diff
@@ -52,6 +53,11 @@ void setUp() throws SQLException {
         connection = extension.getDatabaseManger().getDataSource().getConnection();
     }
 
+    @AfterEach
+    void tearDown() {
+        extension.getDatabaseManger().shutdown();
```
Comment (AndrewBoyarsky) [inline apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/db/ShardRecoveryDaoJdbcTest.java]: Extension support automatic shutdown

File: apl-core/src/test/java/com/apollocurrency/aplwallet/apl/testutil/DbManipulator.java [hunk 12]
```diff
@@ -32,8 +32,17 @@ public DbManipulator(Path dbFile) {
 
     public DbManipulator(DbProperties dbProperties, PropertiesHolder propertiesHolder) {
         Objects.requireNonNull(dbProperties, "dbProperties is NULL");
-        this.databaseManager = new DatabaseManagerImpl(dbProperties, propertiesHolder == null ? new PropertiesHolder() : propertiesHolder);
-        this.populator = new DbPopulator(databaseManager.getDataSource(), "db/schema.sql", "db/data.sql");
+        PropertiesHolder propertiesHolderParam = propertiesHolder == null ? new PropertiesHolder() : propertiesHolder;
+        this.databaseManager = new DatabaseManagerImpl(dbProperties, propertiesHolderParam);
+
+        String dataScriptPath = "db/data.sql";
```
Comment (AndrewBoyarsky) [inline apl-core/src/test/java/com/apollocurrency/aplwallet/apl/testutil/DbManipulator.java]: Why we require such code here?

File: apl-core/src/test/resources/db/data.sql [hunk 13]
```diff
@@ -74,10 +74,10 @@ INSERT into Public.TRANSACTION_SHARD_INDEX(transaction_id, partial_transaction_h
 (102,X'b96d5e9f64e51c597513717691eeeeaf18a26a864034f62c' ,1),
 (103,X'cca5a1f825f9b918be00f35406f70b108b6656b299755558' ,1)
 ;
-INSERT into PUBLIC.SHARD (shard_id, shard_hash) VALUES
-(1, X'8dd2cb2fcd453c53b3fe53790ac1c104a6a31583e75972ff62bced9047a15176'),
-(2, X'a3015d38155ea3fd95fe8952f579791e4ce7f5e1e21b4ca4e0c490553d94fb7d'),
-(3, X'931A8011F4BA1CDC0BCAE807032FE18B1E4F0B634F8DA6016E421D06C7E13693')
+INSERT into PUBLIC.SHARD (shard_id, shard_hash, shard_height) VALUES
+(1, X'8dd2cb2fcd453c53b3fe53790ac1c104a6a31583e75972ff62bced9047a15176', 1),
```
Comment (AndrewBoyarsky) [inline apl-core/src/test/resources/db/data.sql]: 1, 2 , 3 - not so good height values for shards. Make it more real.

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/AplDbVersion.java (hunk 1 above)
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/AplDbVersion.java]: why should ? Why we can have it as Long for limited number of rows?

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/AplDbVersion.java (hunk 3 above)
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/AplDbVersion.java]: That variant is more safe, if it's needed.

File: apl-core/src/test/java/com/apollocurrency/aplwallet/apl/testutil/DbManipulator.java (hunk 12 above)
Comment (yuriy-larin-firstbridge) [reply apl-core/src/test/java/com/apollocurrency/aplwallet/apl/testutil/DbManipulator.java]: It's possible to skip test data for manual testing in 'spacial case' when 'skip parameter' is specified explicitly.

File: apl-core/src/test/resources/db/data.sql (hunk 13 above)
Comment (yuriy-larin-firstbridge) [reply apl-core/src/test/resources/db/data.sql]: It's a fake shards, we don't need more real values here. It works quite good now.

File: apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/db/ShardRecoveryDaoJdbcTest.java (hunk 11 above)
Comment (yuriy-larin-firstbridge) [reply apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/db/ShardRecoveryDaoJdbcTest.java]: Done

File: apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/db/BlockDaoTest.java (hunk 10 above)
Comment (yuriy-larin-firstbridge) [reply apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/db/BlockDaoTest.java]: Done

File: apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/db/ShardDataSourceCreateHelperTest.java (hunk 9 above)
Comment (yuriy-larin-firstbridge) [reply apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/db/ShardDataSourceCreateHelperTest.java]: Done

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/DataTransferManagementReceiverImpl.java (hunk 5 above)
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/DataTransferManagementReceiverImpl.java]: Done

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/ShardDataSourceCreateHelper.java (hunk 4 above)
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/ShardDataSourceCreateHelper.java]: Done

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/AplDbVersion.java (hunk 2 above)
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/AplDbVersion.java]: Removed

File: apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/db/BlockDaoTest.java (hunk 7 above)
Comment (yuriy-larin-firstbridge) [reply apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/db/BlockDaoTest.java]: I copied it from another test

File: apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/db/BlockDaoTest.java (hunk 8 above)
Comment (yuriy-larin-firstbridge) [reply apl-core/src/test/java/com/apollocurrency/aplwallet/apl/core/db/BlockDaoTest.java]: Done

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/AplDbVersion.java (hunk 1 above)
Comment (AndrewBoyarsky) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/AplDbVersion.java]: Look for height type in other tables and you will understand. Long type is misleading

File: apl-core/src/test/java/com/apollocurrency/aplwallet/apl/testutil/DbManipulator.java (hunk 12 above)
Comment (AndrewBoyarsky) [reply apl-core/src/test/java/com/apollocurrency/aplwallet/apl/testutil/DbManipulator.java]: Please remove this. Simply add option to Constructor `schemaOnly` if required.

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/AplDbVersion.java (hunk 1 above)
Comment (yuriy-larin-firstbridge) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/AplDbVersion.java]: Replaced by Int - Integer

File: apl-core/src/test/java/com/apollocurrency/aplwallet/apl/testutil/DbManipulator.java (hunk 12 above)
Comment (yuriy-larin-firstbridge) [reply apl-core/src/test/java/com/apollocurrency/aplwallet/apl/testutil/DbManipulator.java]: I didn't want change current behavior, so it works as before. The **data.sql** is loaded by default.
It's possible to do not load it by supplying "apl.testData = false" explicitly when PropertiesHolder exists and contains it.

**Pull request merged**